### ISSUE TITLE
More completions

### ISSUE THREAD

**Issue by The-Compiler** [qutebrowser/qutebrowser#32]
More things should be supported for the completion, for example:
- [x] URLs (history)
- [ ] Flags for commands
- [ ] Choices for commands
- [ ] Choices for flags with arguments (#1532)
- [ ] Userscript in userscript paths with `:spawn --userscript` (#1543)
- [ ] Keys for `:bind`
- [x] Commands for `:bind`, aliases, etc.
- [ ] Searches
- [x] Tabs (with a `:buffer` command)
- [x] Tabs in current window for `:tab-focus`
- [ ] Aliased commands
- [ ] Commands after `;;`
- [ ] File paths for download, etc. (also maybe for `:open` after entering `/` or `file://`)
- [x] Closed tabs for `:undo`
- [x] back/forward history for `:back`/`:forward` (#1561)
- [x] Search engine names (PR at #2191, after SQL completion)
- [ ] Search engine results/suggestions?
- [ ] Downloads for `download-{open,delete,remove,retry}`
- [ ] Commands from command history
- [ ] Existing quickmark names when saving a quickmark
- [ ] Elements for list/dict settings
- [ ] Anchors for `:scroll-to-anchor`
- [ ] Available fonts for font settings
- [ ] Commands and their arguments with `:later` / `:repeat` / `:run-with-count` (see comment below)

**Comment by lahwaacz** [comment]
If I may add something to the wishlist...

- search engines keywords after `:open`
- command history

**Comment by The-Compiler** [comment]
`:buffer` was implemented by @toofar in #1317 :tada:

**Comment by rcorre** [comment]
For `back/forward`, I'm imagining functionality similar to `buffer`

**Comment by rcorre** [comment]
Edited to add `download-*` completion.

**Comment by The-Compiler** [comment]
#2191 has some code for searchengine name completions - however, we'll need to revisit that after SQL completion (#2295) is in.

For posterity:

```diff
From 3033cc7e1cc1f97878fb17905f1e50e316717817 Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Sun, 25 Dec 2016 20:07:35 +0100
Subject: [PATCH 1/8] added simple searchengine completion to
 UrlCompletionModel

---
 qutebrowser/completion/models/urlmodel.py | 12 ++++++++++++
 1 file changed, 12 insertions(+)

diff --git a/qutebrowser/completion/models/urlmodel.py b/qutebrowser/completion/models/urlmodel.py
index a222df4..60c3a22 100644
--- a/qutebrowser/completion/models/urlmodel.py
+++ b/qutebrowser/completion/models/urlmodel.py
@@ -50,6 +50,7 @@ def __init__(self, parent=None):
         self._quickmark_cat = self.new_category("Quickmarks")
         self._bookmark_cat = self.new_category("Bookmarks")
         self._history_cat = self.new_category("History")
+        self._searchengine_cat = self.new_category("Searchengines")
 
         quickmark_manager = objreg.get('quickmark-manager')
         quickmarks = quickmark_manager.marks.items()
@@ -76,6 +77,10 @@ def __init__(self, parent=None):
         self._history.add_completion_item.connect(self.on_history_item_added)
         self._history.cleared.connect(self.on_history_cleared)
 
+        for se_name, se_url in config.section('searchengines').items():
+            self.new_item(self._searchengine_cat, se_name, se_url)
+        objreg.get('config').changed.connect(self.on_searchengines_changed)
+
         objreg.get('config').changed.connect(self.reformat_timestamps)
 
     def _fmt_atime(self, atime):
@@ -190,3 +195,10 @@ def delete_cur_item(self, completion):
             qtutils.ensure_valid(sibling)
             name = sibling.data()
             quickmark_manager.delete(name)
+
+    @config.change_filter('searchengines')
+    def on_searchengines_changed(self):
+        """Reload searchengines if the config section was changed."""
+        self._searchengine_cat.removeRows(0, self._searchengine_cat.rowCount())
+        for se_name, se_url in config.section('searchengines').items():
+            self.new_item(self._searchengine_cat, se_name, se_url)

From f5c05e705e5a4ac2c3688d52867c0e18cddc6a87 Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Sun, 25 Dec 2016 21:59:42 +0100
Subject: [PATCH 2/8] added searchengines to test_url_completion

---
 tests/unit/completion/test_models.py | 23 +++++++++++++++++++----
 1 file changed, 19 insertions(+), 4 deletions(-)

diff --git a/tests/unit/completion/test_models.py b/tests/unit/completion/test_models.py
index 84c87b2..9c770a5 100644
--- a/tests/unit/completion/test_models.py
+++ b/tests/unit/completion/test_models.py
@@ -167,6 +167,16 @@ def web_history(stubs, web_history_stub):
     return web_history_stub
 
 
+@pytest.fixture
+def searchengines(config_stub):
+    """Pre-populate the config with some searchentries."""
+    config_stub.data['searchengines'] = collections.OrderedDict([
+        ('test', 'http://www.example.org/?q={}'),
+        ('test-with-dash', 'http://www.example.org/?q={}'),
+    ])
+    return config_stub
+
+
 def test_command_completion(qtmodeltester, monkeypatch, stubs, config_stub,
                             key_config_stub):
     """Test the results of command completion.
@@ -264,17 +274,18 @@ def test_bookmark_completion(qtmodeltester, bookmarks):
 
 
 def test_url_completion(qtmodeltester, config_stub, web_history, quickmarks,
-                        bookmarks):
+                        bookmarks, searchengines):
     """Test the results of url completion.
 
     Verify that:
-        - quickmarks, bookmarks, and urls are included
+        - quickmarks, bookmarks, searchengines and urls are included
         - no more than 'web-history-max-items' history entries are included
         - the most recent entries are included
     """
     config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
                                       'web-history-max-items': 2}
     model = urlmodel.UrlCompletionModel()
+
     qtmodeltester.data_display_may_return_none = True
     qtmodeltester.check(model)
 
@@ -293,12 +304,16 @@ def test_url_completion(qtmodeltester, config_stub, web_history, quickmarks,
             ('https://python.org', 'Welcome to Python.org', '2016-03-08'),
             ('https://github.com', 'GitHub', '2016-05-01'),
         ],
+        "Searchengines": [
+            ('test', 'http://www.example.org/?q={}', ''),
+            ('test-with-dash', 'http://www.example.org/?q={}', ''),
+        ],
     })
 
 
 def test_url_completion_delete_bookmark(qtmodeltester, config_stub,
                                         web_history, quickmarks, bookmarks,
-                                        qtbot):
+                                        searchengines, qtbot):
     """Test deleting a bookmark from the url completion model."""
     config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
                                       'web-history-max-items': 2}
@@ -316,7 +331,7 @@ def test_url_completion_delete_bookmark(qtmodeltester, config_stub,
 
 def test_url_completion_delete_quickmark(qtmodeltester, config_stub,
                                          web_history, quickmarks, bookmarks,
-                                         qtbot):
+                                         searchengines, qtbot):
     """Test deleting a bookmark from the url completion model."""
     config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
                                       'web-history-max-items': 2}

From 5fb986db6c6d5c135129416633e5f92b5b6e6ef8 Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Sun, 25 Dec 2016 22:34:03 +0100
Subject: [PATCH 3/8] sorted searchengines for completion

---
 qutebrowser/completion/models/urlmodel.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/qutebrowser/completion/models/urlmodel.py b/qutebrowser/completion/models/urlmodel.py
index 60c3a22..f99437d 100644
--- a/qutebrowser/completion/models/urlmodel.py
+++ b/qutebrowser/completion/models/urlmodel.py
@@ -77,7 +77,7 @@ def __init__(self, parent=None):
         self._history.add_completion_item.connect(self.on_history_item_added)
         self._history.cleared.connect(self.on_history_cleared)
 
-        for se_name, se_url in config.section('searchengines').items():
+        for se_name, se_url in sorted(config.section('searchengines').items()):
             self.new_item(self._searchengine_cat, se_name, se_url)
         objreg.get('config').changed.connect(self.on_searchengines_changed)
 
@@ -200,5 +200,5 @@ def delete_cur_item(self, completion):
     def on_searchengines_changed(self):
         """Reload searchengines if the config section was changed."""
         self._searchengine_cat.removeRows(0, self._searchengine_cat.rowCount())
-        for se_name, se_url in config.section('searchengines').items():
+        for se_name, se_url in sorted(config.section('searchengines').items()):
             self.new_item(self._searchengine_cat, se_name, se_url)

From 86c2e619f992d76989a264d53c58dd9f7b215c7b Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Sun, 25 Dec 2016 23:27:39 +0100
Subject: [PATCH 4/8] added test for modified searchengine config

---
 tests/unit/completion/test_models.py | 34 ++++++++++++++++++++++++++++++++++
 1 file changed, 34 insertions(+)

diff --git a/tests/unit/completion/test_models.py b/tests/unit/completion/test_models.py
index 9c770a5..f540b33 100644
--- a/tests/unit/completion/test_models.py
+++ b/tests/unit/completion/test_models.py
@@ -310,6 +310,40 @@ def test_url_completion(qtmodeltester, config_stub, web_history, quickmarks,
         ],
     })
 
+def test_url_completion_add_searchengine(qtmodeltester, config_stub, web_history, quickmarks,
+                        bookmarks, searchengines):
+    """Test adding searchengine to config."""
+    config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
+                                      'web-history-max-items': 2}
+
+    model = urlmodel.UrlCompletionModel()
+    
+    qtmodeltester.data_display_may_return_none = True
+    qtmodeltester.check(model)
+
+    config_stub.set('searchengines', 'newEntry', 'http://www.example.org/?q={}')
+    _check_completions(model, {
+        "Quickmarks": [
+            ('https://wiki.archlinux.org', 'aw', ''),
+            ('https://duckduckgo.com', 'ddg', ''),
+            ('https://wikipedia.org', 'wiki', ''),
+        ],
+        "Bookmarks": [
+            ('https://github.com', 'GitHub', ''),
+            ('https://python.org', 'Welcome to Python.org', ''),
+            ('http://qutebrowser.org', 'qutebrowser | qutebrowser', ''),
+        ],
+        "History": [
+            ('https://python.org', 'Welcome to Python.org', '2016-03-08'),
+            ('https://github.com', 'GitHub', '2016-05-01'),
+        ],
+        "Searchengines": [
+            ('test', 'http://www.example.org/?q={}', ''),
+            ('test-with-dash', 'http://www.example.org/?q={}', ''),
+            ('newEntry', 'http://www.example.org/?q={}', ''),
+        ],
+    })
+
 
 def test_url_completion_delete_bookmark(qtmodeltester, config_stub,
                                         web_history, quickmarks, bookmarks,

From 92c6286f4b1c49933407d7a1715e7bd861b1f39b Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Mon, 26 Dec 2016 00:12:59 +0100
Subject: [PATCH 5/8] fixed lint warnings in test_models.py

---
 tests/unit/completion/test_models.py | 10 ++++++----
 1 file changed, 6 insertions(+), 4 deletions(-)

diff --git a/tests/unit/completion/test_models.py b/tests/unit/completion/test_models.py
index f540b33..07d9195 100644
--- a/tests/unit/completion/test_models.py
+++ b/tests/unit/completion/test_models.py
@@ -310,18 +310,20 @@ def test_url_completion(qtmodeltester, config_stub, web_history, quickmarks,
         ],
     })
 
-def test_url_completion_add_searchengine(qtmodeltester, config_stub, web_history, quickmarks,
-                        bookmarks, searchengines):
+def test_url_completion_add_searchengine(qtmodeltester, config_stub,
+                                         web_history, quickmarks, bookmarks,
+                                         searchengines):
     """Test adding searchengine to config."""
     config_stub.data['completion'] = {'timestamp-format': '%Y-%m-%d',
                                       'web-history-max-items': 2}
 
     model = urlmodel.UrlCompletionModel()
-    
+
     qtmodeltester.data_display_may_return_none = True
     qtmodeltester.check(model)
 
-    config_stub.set('searchengines', 'newEntry', 'http://www.example.org/?q={}')
+    config_stub.set('searchengines', 'newEntry',
+                    'http://www.example.org/?q={}')
     _check_completions(model, {
         "Quickmarks": [
             ('https://wiki.archlinux.org', 'aw', ''),

From da575d1e6ab18494219e7c884132e487c75a0d73 Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Mon, 26 Dec 2016 00:52:53 +0100
Subject: [PATCH 6/8] added missing line

---
 tests/unit/completion/test_models.py | 1 +
 1 file changed, 1 insertion(+)

diff --git a/tests/unit/completion/test_models.py b/tests/unit/completion/test_models.py
index 07d9195..408ba28 100644
--- a/tests/unit/completion/test_models.py
+++ b/tests/unit/completion/test_models.py
@@ -310,6 +310,7 @@ def test_url_completion(qtmodeltester, config_stub, web_history, quickmarks,
         ],
     })
 
+
 def test_url_completion_add_searchengine(qtmodeltester, config_stub,
                                          web_history, quickmarks, bookmarks,
                                          searchengines):

From aa75b1f413a10148dd35606585354ba9c0f52eaf Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Sun, 5 Feb 2017 18:05:23 +0100
Subject: [PATCH 7/8] appended space to se_name

---
 qutebrowser/completion/models/urlmodel.py | 4 ++--
 1 file changed, 2 insertions(+), 2 deletions(-)

diff --git a/qutebrowser/completion/models/urlmodel.py b/qutebrowser/completion/models/urlmodel.py
index f99437d..e87f224 100644
--- a/qutebrowser/completion/models/urlmodel.py
+++ b/qutebrowser/completion/models/urlmodel.py
@@ -78,7 +78,7 @@ def __init__(self, parent=None):
         self._history.cleared.connect(self.on_history_cleared)
 
         for se_name, se_url in sorted(config.section('searchengines').items()):
-            self.new_item(self._searchengine_cat, se_name, se_url)
+            self.new_item(self._searchengine_cat, se_name + ' ', se_url)
         objreg.get('config').changed.connect(self.on_searchengines_changed)
 
         objreg.get('config').changed.connect(self.reformat_timestamps)
@@ -201,4 +201,4 @@ def on_searchengines_changed(self):
         """Reload searchengines if the config section was changed."""
         self._searchengine_cat.removeRows(0, self._searchengine_cat.rowCount())
         for se_name, se_url in sorted(config.section('searchengines').items()):
-            self.new_item(self._searchengine_cat, se_name, se_url)
+            self.new_item(self._searchengine_cat, se_name + ' ', se_url)

From 85484747e9dd2f8e1553e8adc2647c941684f049 Mon Sep 17 00:00:00 2001
From: rsteube <rsteube@users.noreply.github.com>
Date: Sun, 5 Feb 2017 19:51:45 +0100
Subject: [PATCH 8/8] updated tests for searchengine completion

---
 tests/unit/completion/test_models.py | 10 +++++-----
 1 file changed, 5 insertions(+), 5 deletions(-)

diff --git a/tests/unit/completion/test_models.py b/tests/unit/completion/test_models.py
index 408ba28..8c8eb06 100644
--- a/tests/unit/completion/test_models.py
+++ b/tests/unit/completion/test_models.py
@@ -305,8 +305,8 @@ def test_url_completion(qtmodeltester, config_stub, web_history, quickmarks,
             ('https://github.com', 'GitHub', '2016-05-01'),
         ],
         "Searchengines": [
-            ('test', 'http://www.example.org/?q={}', ''),
-            ('test-with-dash', 'http://www.example.org/?q={}', ''),
+            ('test ', 'http://www.example.org/?q={}', ''),
+            ('test-with-dash ', 'http://www.example.org/?q={}', ''),
         ],
     })
 
@@ -341,9 +341,9 @@ def test_url_completion_add_searchengine(qtmodeltester, config_stub,
             ('https://github.com', 'GitHub', '2016-05-01'),
         ],
         "Searchengines": [
-            ('test', 'http://www.example.org/?q={}', ''),
-            ('test-with-dash', 'http://www.example.org/?q={}', ''),
-            ('newEntry', 'http://www.example.org/?q={}', ''),
+            ('test ', 'http://www.example.org/?q={}', ''),
+            ('test-with-dash ', 'http://www.example.org/?q={}', ''),
+            ('newEntry ', 'http://www.example.org/?q={}', ''),
         ],
     })
 ```

**Comment by The-Compiler** [comment]
Some of the completions are low-hanging fruits now that #2295 is in.

**Comment by Enzime** [comment]
I think we should add support for `:tab-*` commands to take `:buffer` like arguments e.g. `:tab-delete`

**Comment by Enzime** [comment]
I think we should add tab completions for `:tab-*` commands so that you could use it like `:tab-delete github`

**Comment by The-Compiler** [comment]
@Enzime What would that do with multiple tabs matching `github`? I think tab completion would be nice, but I'm not sure about the fuzzy-matching that `:buffer` does currently.

**Comment by Enzime** [comment]
If it matches more than one tab, it could either error out (unless a flag to delete multiple is given), alternatively we could make it prompt for which tab to delete.

One way we could prompt for which tab to delete would be by not allowing the user to hit Enter when it matches more than one tab and pop up a message each time they try to press Enter.

**Comment by pezcore** [comment]
We do need a tab-hsistory dialog. Without the ability to view a list of the indices of back and foward like in firefox when you right click the back button :

**Comment by The-Compiler** [comment]
@pezcore That's why there is *"back/forward history for `:back`/`:forward`"* above.

**Comment by The-Compiler** [comment]
Searchengine completion just landed via #4004 - thanks @rsteube for the original PR and @philipcs for tackling this with the updated completion code.

**Comment by The-Compiler** [comment]
Completion for `:back`, `:forward` and `:undo` just landed FWIW!

**Comment by The-Compiler** [comment]
Another one: Commands and their arguments with `:later` / `:repeat` / `:run-with-count`

This is only a partial solution:

```diff
diff --git i/qutebrowser/completion/models/miscmodels.py w/qutebrowser/completion/models/miscmodels.py
index 925f95bbb..4f44b88e6 100644
--- i/qutebrowser/completion/models/miscmodels.py
+++ w/qutebrowser/completion/models/miscmodels.py
@@ -28,7 +28,7 @@
 from qutebrowser.browser import inspector
 
 
-def command(*, info):
+def command(*_args, info):
     """A CompletionModel filled with non-hidden commands and descriptions."""
     model = completionmodel.CompletionModel(column_widths=(20, 60, 20))
     cmdlist = util.get_cmd_completions(info, include_aliases=True,
diff --git i/qutebrowser/misc/utilcmds.py w/qutebrowser/misc/utilcmds.py
index 56138c798..d8d6b3a4a 100644
--- i/qutebrowser/misc/utilcmds.py
+++ w/qutebrowser/misc/utilcmds.py
@@ -36,11 +36,13 @@
 from qutebrowser.api import cmdutils
 from qutebrowser.misc import (  # pylint: disable=unused-import
     consolewidget, debugcachestats, objects, miscwidgets)
+from qutebrowser.completion.models import miscmodels
 from qutebrowser.utils.version import pastebin_version
 from qutebrowser.qt import sip
 
 
 @cmdutils.register(maxsplit=1, no_cmd_split=True, no_replace_variables=True)
+@cmdutils.argument('command', completion=miscmodels.command)
 @cmdutils.argument('win_id', value=cmdutils.Value.win_id)
 def later(ms: int, command: str, win_id: int) -> None:
     """Execute a command after some time.
@@ -70,6 +72,7 @@ def later(ms: int, command: str, win_id: int) -> None:
 
 
 @cmdutils.register(maxsplit=1, no_cmd_split=True, no_replace_variables=True)
+@cmdutils.argument('command', completion=miscmodels.command)
 @cmdutils.argument('win_id', value=cmdutils.Value.win_id)
 @cmdutils.argument('count', value=cmdutils.Value.count)
 def repeat(times: int, command: str, win_id: int, count: int = None) -> None:
@@ -91,6 +94,7 @@ def repeat(times: int, command: str, win_id: int, count: int = None) -> None:
 
 
 @cmdutils.register(maxsplit=1, no_cmd_split=True, no_replace_variables=True)
+@cmdutils.argument('command', completion=miscmodels.command)
 @cmdutils.argument('win_id', value=cmdutils.Value.win_id)
 @cmdutils.argument('count', value=cmdutils.Value.count)
 def run_with_count(count_arg: int, command: str, win_id: int,

```

as it continues to complete command names after a space is entered.

**Comment by MrbeanN513** [comment]
how to disable automatic pop up tab on Qutebrowser??

**Comment by The-Compiler** [comment]
@MrbeanN513 This seems totally unrelated to this issue - please [open a new discussion](https://github.com/qutebrowser/qutebrowser/discussions/new) and explain in more detail what you mean with "automatic pop up tab".

**Comment by samyak-jain** [comment]
Does #2191 tackle search engine suggestions? Or is there a separate PR/issue for that?

**Comment by The-Compiler** [comment]
@samyak-jain I'm not sure what you're asking. #2191 wasn't merged, but #4004 adds search engine names to the completion. That's a different thing that search term suggestions though (see the *"Search engine results/suggestions?"* point above).

**Comment by samyak-jain** [comment]
@The-Compiler Gotcha. So there's no work started on Search results/suggestions? I can give it a shot if that's ok.

**Comment by The-Compiler** [comment]
No, and I'm not sure if it's even possible to implement that with the current way the completion works - it's definitely not something I'd recommend as a first contribution, I'm afraid.

**Comment by samyak-jain** [comment]
@The-Compiler If it's not too much to ask, do you mind explaining at a high level, what it would take to develop this/what is blocking this? Feel free to link me if you've discussed this somewhere else earlier.

**Comment by The-Compiler** [comment]
I won't have the time to dig deeper into this at the moment, but from the back of my head:

- If we have a search engine URL in the config, how are we going to get autocompletion for that? Probably via OpenSearch (#717), but I'm not sure if there even is a standardized way?
- If there are APIs from search engines, are we allowed to use them? Do we need API keys or something?
- How to do this in a privacy preserving way? We want to avoid sending things to third-parties whenever possible.
- How to do this in a performant way? Currently there's no way for completions to be updated "on the fly" (i.e. the entire completion needs to be generated before it is shown), and doing blocking HTTP requests (or even delaying showing the completion before we get search engine results) probably isn't acceptable.
- Probably more I'm not thinking of at the moment.

Right now, I'm treating this as a "wishlist" item - if someone can reasonably convince me that it's actually worth the complexity (or not as complex as I think after all) I might accept a contribution, but it's quite debatable whether it's worth doing it all - and either way, there still are a lot of unknowns.

**Comment by samyak-jain** [comment]
Thanks for your comments. Here are some of my thoughts. 

- I think that's a good idea. I've looked at a few search engines and using OpenSearch discovery seems to work well in figuring out the auto complete url. We could maybe hard code some of the popular search engines and allow users to enter their own search engines where we can autodiscover using OpenSearch. This can simplify adding search engines because we can also get the URL to search with, programatically rather than asking users to give a template url which is the way things are currently done. 
- It seems to be working without any API Keys for now. Here is a sample for google: https://google.com/complete/search?client=firefox&q=test and a sample for duckduckgo: https://duckduckgo.com/ac/?q=test&type=list. The format across search engines seems fairly consistent. Regarding if we are allowed to use them. So far, there don't seem to be any issues. Of course, they can close these for use anytime but I don't see this getting closed as long as open source browsers like Chromium have these features (since they can always be reverse engineered by someone else). I agree that this part is out of our control and the best we can hope for is that search engines don't close this. This blog by google: https://developers.google.com/search/blog/2015/07/update-on-autocomplete-api seems to suggest that they are at least aware that people are doing this. They did say they would close this feature in 2015 but it has survived thus far. Even if google closes this, I suspect most other search engines wouldn't follow through. 
- There's no avoiding 3rd party requests of course. I think the best course of action is to make this opt-in and maybe make it clear in the documentation that enabling this option would require us to send requests to 3rd party services.
- This would certainly be a blocker because suggestions would change with each character the user types. We need the ability to asynchrously send requests and update the ui in a non blocking way. Would love to get your inputs on this. Can we run this on QThreads and update the UI? If the implementation is too complicated, are there any other issues/features that can benefit from adding this? Just to make this more "worth it". 

I personally don't mind giving this a shot if you feel this is feasbile though I totally understand if you don't have the bandwidth to review features that are not a priority. The way I see it, it certainly requires some bit of effort but I think it may be worth doing this since it requires making improvements to other components as well. 
Interested to know if you think this is worth pursuing or if there are any additional concerns.

**Comment by andrewcarlotti** [comment]
> * How to do this in a performant way? Currently there's no way for completions to be updated "on the fly" (i.e. the entire completion needs to be generated before it is shown), and doing blocking HTTP requests (or even delaying showing the completion before we get search engine results) probably isn't acceptable.

I think it would be worth resolving this anyway, since this is currently an issue with browser history completion. If I try opening a page that triggers searching my entire browser history from disk while I'm watching a video, then the video graphics stop updating and the audio stops a few second later (playback resumes once the history search is complete).

**Comment by The-Compiler** [comment]
> I won't have the time to dig deeper into this at the moment, but from the back of my head:
> 
> * If we have a search engine URL in the config, how are we going to get autocompletion for that? Probably via OpenSearch (#717), but I'm not sure if there even is a standardized way?
> * If there are APIs from search engines, are we allowed to use them? Do we need API keys or something?
> * How to do this in a privacy preserving way? We want to avoid sending things to third-parties whenever possible.
> * How to do this in a performant way? Currently there's no way for completions to be updated "on the fly" (i.e. the entire completion needs to be generated before it is shown), and doing blocking HTTP requests (or even delaying showing the completion before we get search engine results) probably isn't acceptable.
> * Probably more I'm not thinking of at the moment.
> 
> Right now, I'm treating this as a "wishlist" item - if someone can reasonably convince me that it's actually worth the complexity (or not as complex as I think after all) I might accept a contribution, but it's quite debatable whether it's worth doing it all - and either way, there still are a lot of unknowns.

**Comment by The-Compiler** [comment]
Whoops, that was the wrong button :laughing: 

Since we now have quite a lot of discussion about one single point here, let's split this off to #6049 instead. I'll answer over there.
